Implement S2 board screen (basic layout and interaction logic)

CloneWith opened the pull request:
This is a placeholder for our target development branch now. Relevant commits would be added here later.

## Tasks

- [x] Basic components
  - [x] Chess piece
  - [x] Instruction display
  - [x] Chess map pool
  - [x] Chat display
  - [x] Virtual board layer
- [x] Chess board logic

CloneWith committed:
Merge remote-tracking branch 'upstream/master' into feature/board-ui

NaughtyChas commented:
This causes code review nightmare, can we focus on local repo changes first then merge in upstream? Or we merge upstream first.

CloneWith committed:
Merge branch 'derived/new' into feature/board-ui

CloneWith commented:
Now we are good.

cdwcgt committed:
remove the thing that maybe no longer be use or re-write

cdwcgt committed:
remove the thing that maybe no longer be use or re-write

cdwcgt committed:
add basic board screen layout

cdwcgt committed:
add basic board screen layout

cdwcgt committed:
introducing constants `board_size`

cdwcgt committed:
introducing constants `board_size`

CloneWith committed:
Remove initial sizing in EmptyBox constructor

CloneWith committed:
Use Vector2.One to set initial size of EmptyBox

NaughtyChas committed:
Merge branch 'feature/mappool-layout' of https://github.com/CloneWith/osu into feature/mappool-layout

CloneWith committed:
Merge branch 'derived/new' into feature/board-ui

CloneWith committed:
Merge pull request #88 from CloneWith/feature/mappool-layout

Base Board Layout

CloneWith committed:
Add new version of board beatmap panel

CloneWith committed:
Refactor InstructionDisplay with GridContainer

CloneWith committed:
Make team and step attributes changeable for InstructionDisplay

CloneWith committed:
Localize instruction strings

CloneWith committed:
Slightly shrink description font size

CloneWith committed:
Add random test step for InstructionDisplay

CloneWith committed:
Don't use rare characters for description string

CloneWith committed:
Reword win step description

Co-authored-by: NaughtyChas <57820700+NaughtyChas@users.noreply.github.com>

CloneWith committed:
Merge branch 'derived/new' into feature/board-ui

CloneWith committed:
Remove currently unused code in BoardScreen

CloneWith committed:
Change `FirstSelected` to a better name

CloneWith committed:
Display board texture when possible in the board area

CloneWith committed:
Add sliding animation for BoardScreen components

CloneWith committed:
Play animation for BoardScreen after new match selected

CloneWith committed:
Merge remote-tracking branch 'upstream/master' into feature/board-ui

CloneWith committed:
Correct XML documents for InstructionsStrings

CloneWith committed:
Add Shiro deployment step to Instructions

CloneWith committed:
Use a list in TournamentMatch to store chess placements

CloneWith committed:
Update border status once on load complete

CloneWith committed:
Fade out top mask when new state is unavailable

CloneWith committed:
Use switch case for team text in BoardBeatmapPanel

CloneWith committed:
Change to switch case again in FumoBeatmapPanel

CloneWith committed:
Finish transforms recursively in `runAnimation()`

CloneWith committed:
Merge remote-tracking branch 'origin/feature/board-ui' into feature/map-display

CloneWith committed:
Merge remote-tracking branch 'origin/feature/board-ui' into feature/instruction-v2

CloneWith committed:
Remove duplicate using

CloneWith committed:
Add shortened team string for instruction use

CloneWith committed:
Texture initialization and nullable fix

CloneWith committed:
Localize OnFire strings

In case we need to customize those!

NaughtyChas committed:
NaughtyChas the Animator

NaughtyChas committed:
Adjust font size for some text

Otherwise your visual will look like padding: 0px;.

NaughtyChas committed:
Add ban pill

NaughtyChas committed:
Animate it!

NaughtyChas committed:
Change anim starting for topMask

NaughtyChas committed:
Add new animation for winning

CloneWith committed:
Merge pull request #97 from CloneWith/feature/instruction-v2

Refactor board instruction display

NaughtyChas committed:
Update osu.Game.Tournament/Components/FumoBeatmapPanel.cs

Co-authored-by: 复予 <clonewith@qq.com>

NaughtyChas committed:
Wrap trophy into a circle

cdwcgt committed:
remove usage of OfType

CloneWith committed:
Remove unnecessary nullable and cast modifiers

CloneWith committed:
Merge branch 'feature/board-ui' into feature/banpill

cdwcgt committed:
passing ChessPlacement to runAnimation

cdwcgt committed:
remove useless FinishiTransform

cdwcgt committed:
remove transform usage here

cdwcgt committed:
remove usage of currentPlacement

CloneWith committed:
Add function to get team strings in TournamentGame

cdwcgt committed:
use GetTeamColour

CloneWith committed:
Merge remote-tracking branch 'origin/feature/banpill' into feature/banpill

# Conflicts:
#	osu.Game.Tournament/Components/FumoBeatmapPanel.cs

cdwcgt committed:
miss-ed

cdwcgt committed:
switch like

CloneWith committed:
Use `GetTeamString` for team strings in InstructionInfo

CloneWith committed:
Add optional colour fallback for GetTeamColour

CloneWith committed:
Use CircularContainer for bottom right pills

CloneWith committed:
Refactor animation functions in FumoBeatmapPanel

CloneWith committed:
Fix full animation playing condition

CloneWith committed:
Merge branch 'derived/new' into feature/board-ui

cdwcgt committed:
cancel the task before new task (updateState)

CloneWith committed:
Merge pull request #98 from CloneWith/feature/banpill

Refine visuals for picked / banned maps

CloneWith committed:
Fix OnFirstSelected patches

Some of them were overwritten by newer commits.

CloneWith committed:
Re-initialize the floating box schedule before calling

CloneWith committed:
Use non-disposing `Clear` methods in MatchCountdown

We are facing a situation which is way too special. Hope the GC would finally clean them up.

CloneWith committed:
Merge pull request #96 from CloneWith/feature/map-display

Add new version of board beatmap panel

CloneWith committed:
Add beatmap ID field for FumoChessPiece

CloneWith committed:
Remove `OnMouseDown` event for FumoChessPiece

This should be handled externally via `Remove()`.

CloneWith committed:
Integrate InstructionDisplay into BoardScreen

CloneWith committed:
Rewrite basic chess placement logic

Some of required features are not implemented, leaving out.

CloneWith committed:
Add method to randomly generate RoundBeatmap

CloneWith committed:
Add ModMapSection and tests (basic implementation)

CloneWith committed:
Add FumoIcon for our customized icons

CloneWith committed:
Display mod icons based on acronyms in ModMapSection

CloneWith committed:
Add remaining map display and subscribe to ladder updates

CloneWith committed:
Add code description for ModMapSection

CloneWith committed:
Also show placeholder when no map available

CloneWith committed:
Add normal mods list in TournamentGame

CloneWith committed:
Add basic implementation of ChessMapPool and add it to BoardScreen

CloneWith committed:
Merge branch 'refs/heads/derived/new' into feature/board-ui

# Conflicts:
#	osu.Game.Tournament/Screens/Countdown/CountdownScreen.cs
#	osu.Game.Tournament/Screens/TournamentScreen.cs
#	osu.Game.Tournament/TournamentSceneManager.cs

CloneWith committed:
Add bottom counter display to ChessMapPool and localize

CloneWith committed:
Fix couplet check algorithm due to ChessPlacement changes

CloneWith committed:
Scale and size adjustments for ChessMapPool

CloneWith committed:
Add virtual board block component

CloneWith committed:
Add virtual board layer to BoardScreen and support resizing

CloneWith committed:
Finish transforms before new animation in DrawableBoardBlock

CloneWith committed:
Pass MouseDown event through FumoChessPiece

CloneWith committed:
Add virtual block-based associations to BoardScreen

CloneWith committed:
Add simple map card selection logic to BoardScreen

Part of the full interaction logic!

CloneWith committed:
Support fetching beatmap information from ladder for FumoChessPiece

CloneWith committed:
Add a single layer for chess pieces in DrawableBoardBlock

CloneWith committed:
Chess placing animation for board

CloneWith committed:
Update board win placement logic

CloneWith committed:
Complete normal action handling of BoardScreen

CloneWith committed:
Clear chess placements record on reset

CloneWith committed:
Remove remaining map indicator in ModMapSection

CloneWith committed:
Fix ModChessCounterText not updating when chess placement changed

CloneWith committed:
Show the team picking the map in beatmap panel instruction

CloneWith committed:
Include default values of board position in JSON

CloneWith committed:
Add constructor with ChessPlacement of FumoChessPiece

This should make life easier in some cases.

CloneWith committed:
M0eBa9aRa1n

CloneWith committed:
Add chess board initialization step

CloneWith committed:
Use CurrentMatch from TournamentMatchScreen

CloneWith committed:
Add winner detection and fix board coordination again

MoeRain the Baka

CloneWith committed:
Update beatmap panel's state only when relevant placement is changed

CloneWith committed:
Winning or on fire is a question hard to answer

CloneWith committed:
Merge remote-tracking branch 'origin/feature/board-ui' into feature/board-ui

CloneWith committed:
Add Shiro chess initialization logic

CloneWith committed:
Introduce reserved beatmap ID value to avoid conflict

CloneWith committed:
Extract a static version of couplet detection for external use

CloneWith committed:
Add shiro chess updating logic

CloneWith committed:
Localize board strings

CloneWith committed:
Explicitly use predefined default beatmap ID for chess placement

CloneWith committed:
Try to fix NVika formatting warning

CloneWith committed:
What can I say

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,467 @@
+            {
+                beatmapInfoContainer.BorderColour = Colour4.White;
+                beatmapInfoContainer.TransformTo(nameof(beatmapInfoContainer.BorderThickness), INNER_BORDER, 500, Easing.OutQuint);
+                backgroundAddition.FlashColour(Color4.White.Opacity(0.9f), 1000, Easing.OutQuint);
```
due to framework issues, consider using a white box as boarder

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,467 @@
+                banPill.FadeOut(300, Easing.OutQuint);
+                trophyIcon.FadeOut(300, Easing.OutQuint);
+                return;
+            }
```
```suggestion
            if (currentMatch.Value == null)
            {
                banPill.FadeOut(300, Easing.OutQuint);
                trophyIcon.FadeOut(300, Easing.OutQuint);
                lastPlacement = null;
                return;
            }
```
if necessary, we may need run `FinishTransforms`?

CloneWith replied:
Yep, I'm going to add it.

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,467 @@
+        {
+            LocalisableString choiceText = TournamentGame.GetTeamString(placement.OwnerTeam, true, @"Map");
+
+            instructText.Font = OsuFont.Torus.With(size: 16, weight: FontWeight.SemiBold);
```
consider put it into switch case?
rather than saving code, it is better to explicitly express the styles of these two states for me.

CloneWith replied:
Cannot really determine this...

This function was introduced because multiple places use the same team strings for displaying. Would using an independent `switch` block break the consistency or make it easier to understand?

cdwcgt replied:
the essence of the question is whether they are unique
at least if not in the switch case, I need to confirm where this font is used
or leave a comment?

CloneWith replied:
After understanding what you are talking about, this should be resolved 😅

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,467 @@
+                    break;
+
+                default:
+                    // don't do anything
```
should we reset banPill and trophyIcon here or early?

CloneWith replied:
The original place should be okay for now, since they would be updated unconditionally whether a status is added or removed.

cdwcgt commented on osu.Game.Tournament/Components/FumoBeatmapPanel.cs:
```diff
@@ -0,0 +1,467 @@
+            scheduledFloatingBoxAnimation?.RunTask();
+            FinishTransforms(true);
+
+            topMask.FadeTo(newPlacement != null && newPlacement.CurrentType != ChoiceType.Pick ? 0.5f : 0,
+                300, Easing.OutQuint);
+            banPill.FadeTo(!playFullAnimation && newPlacement?.CurrentType == ChoiceType.Ban ? 1 : 0, 300, Easing.OutQuint);
+            trophyIcon.FadeTo(!playFullAnimation && newPlacement?.CurrentType is ChoiceType.RedWin or ChoiceType.BlueWin ? 1 : 0,
+                300, Easing.OutQuint);
```
put these under `if (newPlacement != null)`?

CloneWith replied:
These lines include cases when other conditions are not met. So we'd better keep them?

cdwcgt commented on osu.Game.Tournament/Screens/Board/BoardScreen.cs:
```diff
@@ -29,50 +33,47 @@ namespace osu.Game.Tournament.Screens.Board
 
-        [Resolved]
-        private TournamentSceneManager? sceneManager { get; set; }
+        // ReSharper disable once CollectionNeverUpdated.Local
```
remove this?

cdwcgt reviewed (changes requested):
only beatmap panel part

CloneWith committed:
Removed unused ReSharper disable annotation

CloneWith committed:
Set `lastPlacement` to null upon match change

Co-authored-by: cdwcgt <loginchen@outlook.com>

CloneWith committed:
Finish existing transforms on match change (to null)

CloneWith committed:
Use bottom box and resizing for smoother "border" animation

CloneWith committed:
Null?!

CloneWith committed:
Add independent Shiro placing button

CloneWith committed:
Add instruction for Shiro placing

CloneWith committed:
Disallow selecting Shiro itself

CloneWith committed:
Null?? null!!

CloneWith committed:
Use inner border constant

cdwcgt commented on osu.Game.Tournament/Screens/Board/BoardScreen.cs:
```diff
@@ -251,216 +282,503 @@ private void load(TextureStore textures)
+                                    switch (pickTeam)
+                                    {
+                                        case TeamColour.None:
+                                            succeeded |= addPlacement(TournamentGame.RESERVED_BEATMAP_ID, block);
```
we need escape Shiro Place mode after place one shiro

CloneWith committed:
Adjust reset button colour and localize dialog strings

CloneWith committed:
Directly use IEnumerable<> for couplet searching

CloneWith committed:
Use return function instead of goto tags

cdwcgt commented on osu.Game.Tournament/Screens/Board/BoardScreen.cs:
```diff
@@ -251,216 +282,503 @@ private void load(TextureStore textures)
+        private bool positionEquals(ChessPlacement? placement, DrawableBoardBlock? block)
+            => placement != null && block != null && placement.BoardRow == block.BoardRow && placement.BoardColumn == block.BoardColumn;
+
+        private void clearShiroSelection()
```
we may want to run this when call `setMode` to clear selection due we no longer in shiro select mode

cdwcgt commented on osu.Game.Tournament/Screens/Board/BoardScreen.cs:
```diff
@@ -471,373 +789,176 @@ private void reset()
-        private List<RoundBeatmap> getMapLine(int startX, int startY, int endX, int endY)
-        {
-            List<RoundBeatmap> mapLine = new List<RoundBeatmap>();
+            bool isCommonType = pickType is RoundStep.Pick or RoundStep.Ban or RoundStep.Win or RoundStep.Shiro;
```
since `pickType == Shiro` will always cause an early return, so we don't need to consider shiro here and below

cdwcgt reviewed (changes requested):
reviewed board screen

cdwcgt commented on osu.Game.Tournament/Screens/Board/Components/ModMapSection.cs:
```diff
@@ -0,0 +1,177 @@
+            updateList();
+        }
+
+        private void placementChanged(object? _, NotifyCollectionChangedEventArgs __)
+            => Scheduler.AddOnce(updateList);
```
don't update list when placement changed
it will re-create FumoBeatmapPanel and cancel the animation

cdwcgt commented on osu.Game.Tournament/Screens/Board/Components/ModChessCounterText.cs:
```diff
@@ -0,0 +1,83 @@
+            Font = OsuFont.Torus.With(weight: FontWeight.SemiBold, size: 18);
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            ladder.CurrentMatch.BindValueChanged(matchChanged);
+            ladder.CurrentMatch.Value?.Round.BindValueChanged(_ => updateCounter());
+            ladder.CurrentMatch.Value?.Round.Value?.Beatmaps.BindCollectionChanged((_, _) => updateCounter());
+
+            if (ladder.CurrentMatch.Value != null)
+                ladder.CurrentMatch.Value.ChessPlacements.CollectionChanged += placementChanged;
+
+            updateCounter();
+        }
```
```suggestion
        protected override void LoadComplete()
        {
            base.LoadComplete();

            ladder.CurrentMatch.BindValueChanged(matchChanged, true);
        }
```
the logic here is repeat with the `matchChanged`
we can call matchChanged directly by set `runOnceImmediately`  as true

cdwcgt replied:
In fact, I would prefer to maintain a bindable currentMatch locally

cdwcgt reviewed (changes requested):
I wish this will be a finally

and it is important to note the `Update Owner` function of the `BoardScreen` control panel needs to be confirmed with the organizer

CloneWith committed:
Allow updating owner of won chess pieces

CloneWith committed:
Change to default pick type after placing one shiro

CloneWith committed:
Remove RoundStep.Shiro detection for common type

CloneWith committed:
Remove unused font update lines

CloneWith committed:
Add missing status and deployment switch handling

CloneWith committed:
Add information area in the control panel

CloneWith committed:
Trigger value change immediately to avoid extra code

Co-authored-by: cdwcgt <loginchen@outlook.com>

CloneWith committed:
Don't update the whole section list upon placement change

CloneWith committed:
Make information placeholder text in one line

NaughtyChas reviewed (commented):
Came across the logic of the program and found out there are many place that needs improvement:

Further bugs and reviewing results will be updated here. Since I believe editing a comment will not result in an *intentional* email spam, I will notify you through @.

1. ~**Except for consumed chesses, every piece on the board can be taken.**~ (Fixed, actualy I simply missed this one))

<details>
<summary>Original content for this issue:</summary>
In section 3.5:

> *赛中任意一方在自己的落子回合中可以牺牲一些自己的赢棋，把棋盘上的Shiro或者**对方的一个赢棋**转变为自己的赢棋。我们称其为“夺棋”*。

The current code only supports taking Shiro on the board, but does not include winning chess on the other team. 

On the UX side, we can make the procedure as follows: 
`selecting consecutive winning chess → detect if next selected chess is on the same team → if no, change the winning status of it`

Users are expected to be able to revert this change by right-clicking the winning chess piece in case they make a mistake on stream.

For your reference, the rules for taking the opposite team's winning chess are:

> 1. 消耗自己的一个三连；
> 2. 消耗自己的一个二连与另外任意两个自己的赢棋。

Only one "chess taking" action is allowed for both teams during the whole gameplay, both winning chess and shiro count. It might not be crucial to add a check on this one, but keep in mind that it will affect the Tie-breaker detection logic, more on this in the next section of the comment.
</details>

---

2. **Tie-breaker Indicator**

Since there is one, we have to make it work as expected. Through testing `TB indicator` will not be lit at any time. 

There are only three rules to determine if the next map will be a tie breaker, in section 3.6:

> *TieBreaker可以通过以下任意一种情况启用。*
>   *1.协商TB。当游戏进行到第13回合时，任意一方可以提出直接进入TB的申请。如果另一方同意，则直接进入TB。如果另一方不同意，则回合继续进行。*
>   *2.死局TB。本场比赛中，倘若接下来无论如何行棋、无论哪一方获胜且无论如何进行夺棋都无法使某一方获胜，则直接进入TB。*
>   *3.停局TB。本场比赛中，倘若棋池内剩下的棋由于棋盘MOD区域的约束条件而无法被落子，则直接进入TB。*

**协商TB**

First detection can be added using a simple counter. When this counter reaches 13, we use *some* ways to remind the streamer and the referees. Each round is counted as:

> *比赛的流程通过回合数来控制。双方BAN棋结束后，开始第1回合，先手方行棋。当图打完并且对该棋子进行了赢棋结算后，本回合结束，回合数加1，进入下一回合，移交行棋权。*

If we implement this check on the lazer, rather than handing this to the referee, please make sure this is implemented as per the rules.

Also, if you're still concerned about what should be displayed on the top left section of the screen, ofc you can add this to it. Helpful for players and sensei.

**死局TB**

It is possible to re-create a scenario where the game should be in TB, but the indicator does not light, under the current ruleset:

<img src="https://github.com/user-attachments/assets/d9786476-3884-40ae-9e2d-2cf008c2e96a" Height=50%>

Well, it should not be a TB theoretically, if neither team has used their chance of chess taking (Under this example, Red used their chance, but Blue did not). However, I believe that this is not considered for the code at this point, so it may count as a bug. 

The board has 16 positions to play on, and available chess pieces are always more than available blocks (On least undernditions, 19 chesses including shiro, 17 chesses considering two bans), so there should be no need to include a check on he number of remaining chess.

I reckon the bug is caused by the consumed chess; otherwise, a direct migration of the TB detection algorithm from the S1 client should be working.

**停局TB**

Referring to the special areas. Currently code will not detect and block chess placements in incorrect positions, which is also optional, but we have to add detection for the rest of the chess still in the chesspool:

|Mod restriction area|Chesses allowed to place|
|:-:|:-:|
|HD Area|NM, FM, Shiro, **HD**|
|HR Area|NM, FM, Shiro, **HR**|
|DT Area|NM, FM, Shiro, **DT**|

You can see Mod icons in the corner of the board PNG. The first four blocks on that side are indicated as the special area.

---
@CloneWith @cdwcgt 

3. **Improvements on display after a winning chess gets taken.**

![image](https://github.com/user-attachments/assets/c38efbb4-0e97-4903-b703-fffaae3e836f)

Please note that winning chess pieces in the bottom left corner of the board are associated with `NM1`, `NM2`, and `HR1`. The only winning chess piece on the blue side is `NM3`. When we execute a chess taking action using these three pieces:

![image](https://github.com/user-attachments/assets/764e3a52-657c-4e29-bebe-a605ec5fc98c)

You will see that `NM1`, `NM2`, and `HR1` have been marked as "Consumed", while `NM3` is labelled as a "Red win." This implementation might work correctly from a coding perspective, but it contradicts the goals of the chess pool. In reality, Red won `NM1`, `NM2`, and `HR1`, while Blue claimed `NM3`. Presenting this information incorrectly could lead to confusion about the statement "Red won NM3" for our viewers.

Although I am uncertain if this issue is worth correcting, not to mention that the pieces on the board are interconnected to the status of the elements in the chess pool, but it is clear that displaying the winning history of each piece inaccurately undermines the purpose of the chess pool. While it may be acceptable to keep the consumed chess pieces labelled as "Consumed," presenting the winning history inaccurately is not a good design choice.

This may cause some mayhem during the fix, but hopefully this issue can be considered carefully.

CloneWith commented:
Nice review on your side. Appending these tasks:

- [x] Specific mod area for virtual board block
- [x] `协商 TB` situation: Round counter

The *bug*s you mentioned in TB detection aren't literal bugs, they are just **not implemented**. So from this you can see how f\*\*ked up (and need to be done) I have been these days.

For the round counter to work properly, automatic round advancing is suggested. Or streamers are needed to adjust the counter via buttons in the control panel. Do we need to prepare both of them?

Anyway, those things would be done in this iteration (till 15th).

---

> The current code only supports taking Shiro on the board, but does not include winning chess on the other team.

This problem should have been resolved in the latest batch of commit. Please check again.

NaughtyChas commented:
> Anyway, those things would be done in this iteration (till 15th).

Bro can take your own time to complete it. Don't stress too much (

As currently most of the features have been completed, and chess pieces can be placed on the board, I think we can submit this version of the code, and it is good for demonstration purposes. As for all the stuff I mentioned, are they OK to be completed at our own pace, I believe?

The only thing I can help with is to keep up with these reviews on more on logic and tournament rules side.

CloneWith commented:
> As currently most of the features have been completed, and chess pieces can be placed on the board, I think we can submit this version of the code, and it is good for demonstration purposes.

Okay, then this PR implements the *first half* of the whole board screen.

NaughtyChas commented:
> For the round counter to work properly, automatic round advancing is suggested. Or streamers are needed to adjust the counter via buttons in the control panel. Do we need to prepare both of them?

We can count a round by setting any chess piece to winning status, or when a Shiro has been deployed.

Theoretically, based on the current tournament rules, undoing a move is not allowed, so the round number can only be incremented. But I suggest decrementing the number of the counter each time a winning chess piece gets removed, or a shiro is removed from the board.

Also, we have to consider if a "chess taking" move will be counted as a round. I think the answer is maybe negative.

Is this the "automatic round advancing" you meant? Offering streamers the ability to change the round number is also a good act, since it can save us from unexpected accidents, which we do not expect to have during the stream.

---

Also, I suggest that we can merge this PR to `derived/new`, as now all basic functions of the gameplay on the board have been finished. We can continue these changes about the TB indicator, rules, restrictions and other stuff in a new PR instead? 

Totally up to you, as if you agree, I will approve this PR after I make further testing and reviewing on the code, focusing more on the basic functions on the board instead.

CloneWith commented:
> Also, I suggest that we can merge this PR to `derived/new`, as now all basic functions of the gameplay on the board have been finished. We can continue these changes about the TB indicator, rules, restrictions and other stuff in a new PR instead?

That's also what I'm considering. We've done too much on this branch.

cdwcgt reviewed (approved):
![takakura_anzu_by_xltxdyka_agadyruaamxscfq](https://github.com/user-attachments/assets/9d73da70-fa97-4963-af0d-eb0e0022afc2)
let go next step

NaughtyChas reviewed (approved):
Same here.

CloneWith commented:
❤️ I'm coming ❤️
